Changing configuration for agent image + registry

donnyyung opened the pull request:
## Description
I'm having trouble runnings tests locally so just PR-ing this so I can see how the tests run in circle, don't review this yet cuz I gotta clean up some of the commits! 

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [ ] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to either submit a docs PR, or tell Matt about the necessary documentation changes.
 - [ ] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.

Commit:
Configure Agent Image + Registry via config.yml

This commit removes the client-side use of the TELEPRESENCE_AGENT_IMAGE
and TELEPRESENCE_REGISTRY environment variables.  Instead we now
configure these values in the config.yml.  We will likely want to move
the rest of the Environment Variables into the config.yml but this is a
good start (since these are the ones people actually set in practice).

Signed-off-by: Donny Yung <donaldyung@datawire.io>

thallgren commented on pkg/install/resource/tm_deployment.go:
```diff
@@ -43,12 +43,13 @@ func (ri *tmDeployment) desiredDeployment(ctx context.Context) *kates.Deployment
 		{Name: "SYSTEMA_PORT", Value: sc.env.SystemAPort},
 		{Name: "CLUSTER_ID", Value: sc.clusterID},
-		{Name: "TELEPRESENCE_REGISTRY", Value: sc.env.Registry},
+		{Name: "TELEPRESENCE_REGISTRY", Value: imgConfig.Registry},
```
This (and the agent image further down) is somewhat problematic from a life-cycle perspective. If the manager is installed and a user's config is modified, then that modification will have no effect because it doesn't cause the manager deployment to change. I know that this problem was already there when modifying the env vars but that was not something that we made apparent to the user the way we do now.

I don't have a good solution for this but I think it's important that there's a caveat and an explanation in the docs.

donnyyung replied:
I think I have tackled this in a future commit by separating the agentImage key into agentImage and webhookAgentImage so it's not one overloaded variable that is easier to distinguish between in the documentation.

Although revisiting this now, I might need to break up imgConfig.Registry into two variables as well.

Anyways, let me know what you think of this solution in the later commits.

thallgren commented on pkg/client/config.go:
```diff
@@ -50,21 +52,26 @@ func (c *Config) UnmarshalYAML(node *yaml.Node) (err error) {
 			if err != nil {
 				return err
 			}
 			continue
```
```suggestion
```

thallgren commented on pkg/client/config.go:
```diff
@@ -50,21 +52,26 @@ func (c *Config) UnmarshalYAML(node *yaml.Node) (err error) {
 			if err != nil {
 				return err
 			}
 			continue
```
```suggestion
```

thallgren commented on pkg/client/config.go:
```diff
@@ -50,21 +52,26 @@ func (c *Config) UnmarshalYAML(node *yaml.Node) (err error) {
+			if err != nil {
+				return err
+			}
+			continue
```
```suggestion
```

thallgren commented on pkg/client/config.go:
```diff
@@ -343,6 +392,10 @@ var defaultConfig = Config{
+	},
+	Images: Images{
+		Registry:   "docker.io/datawire",
+		AgentImage: "",
```
Shouldn't these defaults be initialized from the current environment variables? I think it's either that or remove them from the env.

donnyyung replied:
Ah yeah sorry about that, I hadn't gotten to the point where I removed the environment variables.  Will be in the next commit.

thallgren commented on pkg/tun/device_darwin.go:
```diff
@@ -143,9 +143,9 @@ type addrIfReq6 struct {
-	preferred      int64 // nolint:structcheck
+	flags          int32 //nolint:structcheck
+	expire         int64 //nolint:structcheck
+	preferred      int64 //nolint:structcheck
```
You can skip those three. This fix is already in `release/v2`.

Commit:
Making tests work for changes

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Lifecycle fix for Webhook Agent Image config

As noted by Thomas, moving these environment variables to the
configuration will make it more apparent how we were using one value for
a lot of things.  With this change, I'm proposing adding an additional
configuration key to separate the AgentImage (a client will use to
specify which image they want to install with their intercept) and
WebhookAgentImage (which is the agent image that the webhook will inject
when it sees the annotation).

This will be easier to document since we can say "if you use
telepresence connect to deploy telepresence, you can set this value to
toggle the image the webhook will inject".

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Add a config key for ignoring login

While we do auto-detect if a cluster is airgapped to determine whether a
user needs to login or not, it's easier for our testing and users to control the behavior
if they can also explicitly tell the cli to ignore login.

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Use config to get registry for regular agent

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Make smoke-tests work w/ config changes

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Add webhookRegistry key

Doing the same thing to registry that I did by splitting up
agentImage and webhookAgentImage. The env var was too bloated in the
past and this seems more clear.

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Bump changelog

Signed-off-by: Donny Yung <donaldyung@datawire.io>

thallgren reviewed (approved):
All looks good.
